core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -381,12 +388,85 @@ private[spark] class TaskSchedulerImpl(
 
   /**
    * Check whether the resources from the WorkerOffer are enough to run at least one task.
+   * Returns None if the resources don't meet the task requirements, otherwise returns
+   * the task resource assignments to give to the next task. Note that the assignments maybe
+   * be empty if no custom resources are used.
    */
-  private def resourcesMeetTaskRequirements(resources: Map[String, Buffer[String]]): Boolean = {
-    val resourcesFree = resources.map(r => r._1 -> r._2.length)
-    val meetsReqs = ResourceUtils.resourcesMeetRequirements(resourcesFree, resourcesReqsPerTask)
-    logDebug(s"Resources meet task requirements is: $meetsReqs")
-    meetsReqs
+  private def resourcesMeetTaskRequirements(
+      taskSet: TaskSetManager,
+      availCpus: Int,
+      availWorkerResources: Map[String, Buffer[String]]
+      ): Option[Map[String, ResourceInformation]] = {
+    val rpId = taskSet.taskSet.resourceProfileId
+    val taskCpus = sc.resourceProfileManager.taskCpusForProfileId(rpId)
+    // check if the ResourceProfile has cpus first since that is common case
+    if (availCpus < taskCpus) return None
+
+    val taskSetProf = sc.resourceProfileManager.resourceProfileFromId(rpId)
+    // remove task cpus since we checked already
+    val tsResources = taskSetProf.taskResources.filterKeys(!_.equals(ResourceProfile.CPUS))
```

Comment:
Thoughts on always keeping cpu resource distinct from other resources in resource profile ? Given number of times this will need to be run, might be worth avoiding ?

Reply:
so I had thought about that and think I even started to code it up, but then it made other things more difficult.  Let me revisit it since that was quite a while ago and things may have changed.

Reply:
Ok so I think this is actually easier now that I added the ResourceProfileBuilder, and like you say I think it makes sense for efficiency.  But unless we want to keep 2 copies of the data, its going to change the public api. I think that is ok, but it means we need to get a bunch of functions like getExecutoreCores, getTaskCpus, etc.  And to do that efficiently it makes sense to pass these in differently to ResourceProfile. I think that is a large enough change it would make sense to do separately if you are ok with that?

Reply:
That would be an api change for resource profile ?
Do we want to do it before this gets released ? (in a subsequent pr is fine)